hugo/content/docs/recipes/validation/dependency-loops.md

```diff
@@ -0,0 +1,258 @@
+
+It is recommended not to implement your own version of that algorithm. Please use an existing solution! The algorithm is able to output every loop with all its members of that loop. Perfect for adding a custom validation!
+
+#### Complex resolution
+
+The directed graph approach can be processed further when there were no loops found:
+
+With a "topological sort" you can get an order that respects all dependencies. Means more or less: You start with the node that has no dependencies, remove it, put it into your sorted list and do the same for the resulting graph again and again until all dependencies were resolved.
+
+The topological sort (as well as the strongly-connected component search) is a standard algorithm in every good graph library.
+
+## How to make it work in Langium?
+
+### Adapt the grammar
+
+We will change the `HelloWorld` grammar from the learning section, so that persons can greet each other. However, we will introduce a validation that "greeting loops" are forbidden.
+
+```langium
+grammar HelloWorld
+
+entry Model:
+    (persons+=Person | greetings+=Greeting)*;
+
+Person:
+    'person' name=ID;
+
+Greeting:
+    greeter=[Person:ID] 'greets' greeted=[Person:ID] '!';
+
+hidden terminal WS: /\s+/;
+terminal ID: /[_a-zA-Z][\w_]*/;
+````
+
+After the change build your grammar with `npm run langium:generate`.
+
+### Loop detection
+
+Now we will add the validation. Here we will use the graph library ‚graphology‘. Please install these two packages (`graphology` contains the data structure, `graphology-components` contains the strongly-connected component search, `graphology-dag` contains the topological sort):
+
+```bash
+npm install graphology graphology-components graphology-dag
+```
+
+Open the `hello-world-validator.ts` and add another validator for `Model`. It is important to say that we do not create a check on the `Greeting` level, because we need the overview over all greetings. The complete overview is given for the `Model` AST node.
+
+```typescript
+const checks: ValidationChecks<HelloWorldAstType> = {
+    Model: validator.checkGreetingCycles, // new!!!
+    Person: validator.checkPersonStartsWithCapital
+};
+```
+
+And here is the implementation:
+
+```typescript
+checkGreetingCycles(model: Model, accept: ValidationAcceptor): void {
+    //arrange the graph
+    const graph = new DirectedGraph<{}, {greeting: Greeting}>();
+    model.persons.forEach(person => {
+        graph.addNode(person.name);
+    })
+    model.greetings.forEach(greeting => {
+        if(greeting.greeter.ref && greeting.greeted.ref && !graph.hasDirectedEdge(greeting.greeter.ref.name, greeting.greeted.ref.name)) {
+            graph.addEdge(greeting.greeter.ref.name, greeting.greeted.ref.name, {
+                greeting
```

Comment:
maybe add a comment in the code, that we need to store the AST node in order to attach a warning to it later?